typed/remove_test.go

```diff
@@ -0,0 +1,646 @@
+	}, {
+		// path to root type
+		//`{"mapOfLists": {"b":["a","c"]}}`,
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists"),
+		),
+		``,
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+	}, {
+		// path to a top-level element
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists", "b"),
+		),
+		`{"mapOfLists": {"d":["e", "f"]}}`,
+		`{"mapOfLists": {"b":["a","c"]}}`,
+	}, {
+		// just one path to leaf element
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists", "b", _V("a")),
+		),
+		`{"mapOfLists":{"b":["c"],"d":["e", "f"]}}`,
+		`{"mapOfLists":{"b":["a"]}}`,
+	}, {
+		// path to non-existant top-level element
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists", "q"),
+		),
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		`{"mapOfLists":null}`,
+	}, {
+		// path with existant top-level but non-existant leaf element
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists", "b", _V("q")),
+		),
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		`{"mapOfLists":{"b":null}}`,
+	}, {
+		// path with existant top-level but non-existant leaf element
+		`{"mapOfLists": {"b":null, "d":["e", "f"]}}`,
+		_NS(
+			_P("mapOfLists", "b"),
+		),
+		`{"mapOfLists": {"d":["e", "f"]}}`,
+		`{"mapOfLists":{"b":null}}`, // same output as previous case, but can be differentiated by input fieldpath.Set
+	}, {
+		// invalid path to just a leaf
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		_NS(
+			_P(_V("a")),
+		),
+		`{"mapOfLists": {"b":["a","c"], "d":["e", "f"]}}`,
+		``,
+	}, {
+		// path to root type
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps"),
+		),
+		``,
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+	}, {
+		// path to a top-level element
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps", "b"),
+		),
+		`{"mapOfMaps": {"d":{"e":"y", "f":"w"}}}`,
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}}}`,
+	}, {
+		// just one path to leaf element
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps", "b", "a"),
+		),
+		`{"mapOfMaps": {"b":{"c":"z"},"d":{"e":"y", "f":"w"}}}`,
+		`{"mapOfMaps": {"b":{"a":"x"}}}`,
+	}, {
+		// path to non-existant top-level element
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps", "q"),
+		),
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		`{"mapOfMaps": null}`,
+	}, {
+		// path with existant top-level but non-existant leaf element
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps", "b", "q"),
+		),
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		`{"mapOfMaps": {"b":null}}`,
```

Comment:
Should we even keep `b` here? Or should we even keep anything at all since the one thing we're supposed to keep doesn't exist?

Reply:
Yea, that's a good question. I guess my thinking for keeping it this way is that the caller can use the fieldpath.Set they passed in to know that the value did not have a path to the element they expected. If we really don't want it to keep b, we could do some post-processing in `ExtractItems` to get rid of it, that way we don't have to modify the walker, because I think doing it in the walker gets tricky.